File: 188-best-time-to-buy-and-sell-stock-iv/188-best-time-to-buy-and-sell-stock-iv.py
class Solution:
    def maxProfit(self, k: int, prices: List[int]) -> int:
        t_cost = [float('inf')] * k
        t_profit = [0] * k
        if k == 0:
            return 0
        for price in prices:
            i = 0
            while(i < k):
                if i == 0:
                    t_cost[i] = min(t_cost[i], price)
                else:
                    t_cost[i] = min(t_cost[i], price - t_profit[i-1])
                t_profit[i] = max(t_profit[i], price - t_cost[i])
                i = i + 1
        return t_profit[-1]

File: 188-best-time-to-buy-and-sell-stock-iv/188-best-time-to-buy-and-sell-stock-iv.java
class Solution {
    public int maxProfit(int k, int[] prices) {
        int N = prices.length;
        int[][] dp = new int[N][k+1];
        if(k == 0)return 0;
        int ans = 0;
        for(int i = 1; i < N; i++)
            for(int j = k-1; j >= 0; j--){
                dp[i][j] = Math.max(dp[i][j], dp[i-1][j]);
                for(int t = 0; t <= i; t++){
                dp[i][j] = Math.max(dp[i][j], dp[i-t][j+1] +prices[i] -prices[i-t]);
                    }
                if(i == N -1){
                    ans = Math.max(dp[i][j], ans);
                }
                
            }
        return ans;
        
    }
}